### PR TITLE
Detach kernel driver to allow claiming of USB interface

### PR DESCRIPTION
**Pull request by MarkMuth**
Hi I'm using a eurolite USB-DMX512-PRO Interface on Ubuntu 18.04 and ran into several problems.

First I had to create a new udev rule so that libusb_open wouldn't fail in LibUsbAdaptor.cpp. It looks like this:

SUBSYSTEM=="usb", ATTR{idVendor}=="04d8", ATTR{idProduct}=="fa63", MODE="0660", GROUP="dialout"

After creating this rule and rebooting /dev/ttyACM0 was created. So some kind of kernel driver was loaded for the eurolite interface.

This however caused libusb_claim_interface to fail. In this pull request I suggest to detach the kernel driver in case of an error and to retry claiming the USB interface.

Ideally the kernel driver should be reattached after closing olad. But that's not done yet. I thought maybe I'm missing something here.

### CODE REVIEW

**Comment by peternewman** [comment]
Hi Mark,

Briefly, our config for various devices is here:
https://wiki.openlighting.org/index.php/OLA_Device_Specific_Configuration

And our udev rules here:
https://github.com/OpenLightingProject/ola/blob/master/debian/ola.udev

**Comment by MarkMuth** [comment]
Hi! Thank you for the links. Using the instructions udev rule + rmmod I can control the interface now with the ola version (0.10.5) shipped with Ubuntu 18.04.

Still I wonder, what is the point about instructing udev to create a device node for my interface if I later on just use rmmod to destroy it? Isn't there a way to instruct udev to ignore my eurolite from the beginning?

**Comment by peternewman** [comment]
So you have to rmmod, even with the udev rule? I'd imagine there may be a way to use udev to tell it not to load a module, but I don't know off the top of my head.

File: libs/usb/LibUsbAdaptor.cpp [hunk 1]
```diff
@@ -80,10 +80,22 @@ bool OpenHandleAndClaimInterface(libusb_device *usb_device,
   }
 
   int r = libusb_claim_interface(*usb_handle, interface);
-  if (r) {
+  if (r == LIBUSB_ERROR_BUSY) {
+	int error = libusb_detach_kernel_driver(*usb_handle, interface);
+	if (error) {
+		OLA_WARN << "Failed to detach kernel driver " << interface
+				 << " on device: " << usb_device;
```
Comment (peternewman) [inline libs/usb/LibUsbAdaptor.cpp]: Can you align the << please

File: libs/usb/LibUsbAdaptor.cpp [hunk 2]
```diff
@@ -80,10 +80,22 @@ bool OpenHandleAndClaimInterface(libusb_device *usb_device,
...
+	} else {
+	  r = libusb_claim_interface(*usb_handle, interface);
+	}
+  }
+
+  if (r)
+  {
```
Comment (peternewman) [inline libs/usb/LibUsbAdaptor.cpp]: Can you move this to the line above please.

File: libs/usb/LibUsbAdaptor.cpp [hunk 3]
```diff
@@ -80,10 +80,22 @@ bool OpenHandleAndClaimInterface(libusb_device *usb_device,
   }
 
   int r = libusb_claim_interface(*usb_handle, interface);
-  if (r) {
+  if (r == LIBUSB_ERROR_BUSY) {
+	int error = libusb_detach_kernel_driver(*usb_handle, interface);
+	if (error) {
+		OLA_WARN << "Failed to detach kernel driver " << interface
```
Comment (peternewman) [inline libs/usb/LibUsbAdaptor.cpp]: Can you change the message to something like "Failed to detach kernal driver for interface" <<

File: libs/usb/LibUsbAdaptor.cpp (hunk 2 above)
Comment (peternewman) [reply libs/usb/LibUsbAdaptor.cpp]: The other tab/space lint errors will also need fixing before merge:
https://travis-ci.org/OpenLightingProject/ola/jobs/404585543#L2174

**Comment by MarkMuth** [comment]
Should be better now ;)

File: libs/usb/LibUsbAdaptor.cpp [hunk 4]
```diff
@@ -80,6 +80,15 @@ bool OpenHandleAndClaimInterface(libusb_device *usb_device,
   }
 
   int r = libusb_claim_interface(*usb_handle, interface);
+  if (r == LIBUSB_ERROR_BUSY) {
+    int error = libusb_detach_kernel_driver(*usb_handle, interface);
+    if (error) {
+         OLA_WARN << "Failed to detach kernel driver for interface " << interface
```
[the remark this answers is not in the document]
Comment (MarkMuth) [reply libs/usb/LibUsbAdaptor.cpp]: Ok, there were too many leading white spaces. Removing them makes the line 79 characters long.

File: libs/usb/LibUsbAdaptor.cpp [hunk 5]
```diff
@@ -80,6 +80,15 @@ bool OpenHandleAndClaimInterface(libusb_device *usb_device,
   }
 
   int r = libusb_claim_interface(*usb_handle, interface);
+  if (r == LIBUSB_ERROR_BUSY) {
+    int error = libusb_detach_kernel_driver(*usb_handle, interface);
```
[the remark this answers is not in the document]
Comment (MarkMuth) [reply libs/usb/LibUsbAdaptor.cpp]: libusb_close() was called after libusb_claim_interface() failed, so I thought it should be called in any error situation.

**Pull request merged**